File: src/lib/tsp.ts
import type { AlgoKey, Matrix, RouteResult } from '../App'

export type GeneticOptions = {
  population?: number
  generations?: number
  mutationRate?: number
  crossoverRate?: number
  eliteFraction?: number
  tournamentSize?: number
}

function totalForOrder(order: number[], m: Matrix) {
  let dist = 0
  let dur = 0
  const legs: RouteResult['legs'] = []
  for (let i = 0; i < order.length - 1; i++) {
    const a = order[i], b = order[i + 1]
    // Validate indices
    if (a === undefined || b === undefined || a < 0 || b < 0 || 
        a >= m.distances.length || b >= m.distances.length ||
        !m.distances[a] || !m.distances[a][b]) {
      console.error('Invalid tour:', order, 'at positions', i, i+1, 'nodes', a, b)
      throw new Error(`Invalid tour: node ${a} or ${b} out of bounds`)
    }
    const d = m.distances[a][b]
    const t = m.durations[a][b]
    dist += d
    dur += t
    legs.push({ from: a, to: b, distance: d, duration: t })
  }
  return { dist, dur, legs }
}

function bruteForce(m: Matrix): RouteResult {
  const n = m.distances.length
  const nodes = Array.from({ length: n - 1 }, (_, i) => i + 1)
  let best: { order: number[]; dist: number; dur: number; legs: RouteResult['legs'] } | null = null

  function permute(arr: number[], l: number) {
    if (l === arr.length) {
      const order = [0, ...arr]
      const t = totalForOrder(order, m)
      if (!best || t.dur < best.dur) best = { order, dist: t.dist, dur: t.dur, legs: t.legs }
      return
    }
    for (let i = l; i < arr.length; i++) {
      ;[arr[l], arr[i]] = [arr[i], arr[l]]
      permute(arr, l + 1)
      ;[arr[l], arr[i]] = [arr[i], arr[l]]
    }
  }
  permute(nodes, 0)
  const b = best!
  return { order: b.order, totalDistance: b.dist, totalDuration: b.dur, legs: b.legs }
}

function nearestNeighbor(m: Matrix): RouteResult {
  const n = m.distances.length
  
  // Validate matrix
  if (!m.distances || !m.durations || n === 0) {
    throw new Error('Invalid matrix provided to nearestNeighbor')
  }
  
  const unvisited = new Set<number>(Array.from({ length: n - 1 }, (_, i) => i + 1))
  const order = [0]
  
  while (unvisited.size) {
    const last = order[order.length - 1]
    let best: { node: number; dur: number } | null = null
    
    for (const u of unvisited) {
      // Validate indices before access
      if (!m.durations[last] || m.durations[last][u] === undefined) {
        console.error('Invalid matrix access:', { last, u, n, matrixSize: m.distances.length })
        throw new Error(`Invalid matrix: missing durations[${last}][${u}]`)
      }
      const d = m.durations[last][u]
      if (!best || d < best.dur) best = { node: u, dur: d }
    }
    
    if (!best) {
      throw new Error('Nearest neighbor failed to find next node')
    }
    
    order.push(best.node)
    unvisited.delete(best.node)
  }
  
  const t = totalForOrder(order, m)
  return { order, totalDistance: t.dist, totalDuration: t.dur, legs: t.legs }
}

function genetic(m: Matrix, opts?: GeneticOptions): RouteResult {
  const n = m.distances.length
  const nodes = Array.from({ length: n - 1 }, (_, i) => i + 1)
  
  // Adaptive parameters based on problem size - more conservative for large problems
  // Reduce population and generations for larger problems to improve speed
  // Defaults (adaptive) can be overridden by opts
  const defaultPOP = n < 20 ? 80 : n < 50 ? 60 : 40
  const defaultGEN = n < 20 ? 150 : n < 50 ? 100 : 50
  const defaultEliteFraction = 0.2
  const defaultTournament = Math.min(5, Math.floor(defaultPOP * 0.15))
  const defaultMut = 0.2
  const defaultCross = 0.8

  const POP = opts?.population ?? defaultPOP
  const GEN = opts?.generations ?? defaultGEN
  const ELITE_SIZE = Math.max(1, Math.floor(POP * (opts?.eliteFraction ?? defaultEliteFraction))) // Keep at least 1
  const TOURNAMENT_SIZE = opts?.tournamentSize ?? Math.max(2, Math.min(POP, defaultTournament))
  const MUT_RATE = opts?.mutationRate ?? defaultMut
  const CROSS_RATE = opts?.crossoverRate ?? defaultCross
  
  function randOrder() {
    const a = nodes.slice()
    for (let i = a.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1))
      ;[a[i], a[j]] = [a[j], a[i]]
    }
    return [0, ...a]
  }
  
  function fitness(order: number[]) {
    try {
      return totalForOrder(order, m).dur
    } catch (e) {
      // Invalid tour, return worst possible fitness
      return Infinity
    }
  }
  
  // Validate a tour
  function isValidTour(order: number[]): boolean {
    if (order.length !== n) return false
    if (order[0] !== 0) return false
    const seen = new Set(order)
    if (seen.size !== n) return false // Duplicates
    for (let i = 0; i < n; i++) {
      if (!seen.has(i)) return false // Missing node
    }
    return true
  }
  
  // Order Crossover (OX) - preserves relative order from parents
  function orderCrossover(parent1: number[], parent2: number[]): number[] {
    const size = parent1.length - 1 // Exclude fixed starting point
    if (size < 2) return parent1.slice() // Too small to crossover
    
    const start = 1 + Math.floor(Math.random() * (size - 1))
    const end = start + 1 + Math.floor(Math.random() * (size - start))
    
    const child = new Array(parent1.length).fill(-1)
    child[0] = 0 // Keep starting point
    
    // Copy segment from parent1
    for (let i = start; i <= end && i < parent1.length; i++) {
      child[i] = parent1[i]
    }
    
    // Fill remaining from parent2 in order
    const usedGenes = new Set(child.filter(g => g !== -1))
    let childIdx = 1
    
    for (let i = 1; i < parent2.length; i++) {
      const gene = parent2[i]
      if (!usedGenes.has(gene)) {
        // Find next empty slot
        while (childIdx < child.length && child[childIdx] !== -1) {
          childIdx++
        }
        if (childIdx < child.length) {
          child[childIdx] = gene
          usedGenes.add(gene)
        }
      }
    }
    
    // Safety check: if any position is still -1, fill with missing nodes
    const allNodes = new Set(Array.from({ length: n }, (_, i) => i))
    for (let i = 1; i < child.length; i++) {
      if (child[i] === -1) {
        // Find a missing node
        for (const node of allNodes) {
          if (!usedGenes.has(node) && node !== 0) {
            child[i] = node
            usedGenes.add(node)
            break
          }
        }
      }
    }
    
    return child
  }
  
  // Inversion Mutation - reverses a random segment (like 2-opt)
  function inversionMutate(order: number[]): number[] {
    if (Math.random() > MUT_RATE) return order
    
    const o = order.slice()
    const i = 1 + Math.floor(Math.random() * (o.length - 2))
    const j = i + 1 + Math.floor(Math.random() * (o.length - i - 1))
    
    // Reverse segment between i and j
    const segment = o.slice(i, j + 1).reverse()
    for (let k = 0; k < segment.length; k++) {
      o[i + k] = segment[k]
    }
    
    return o
  }
  
  // Swap Mutation - swaps two random positions
  function swapMutate(order: number[]): number[] {
    if (Math.random() > MUT_RATE) return order
    
    const o = order.slice()
    const i = 1 + Math.floor(Math.random() * (o.length - 1))
    const j = 1 + Math.floor(Math.random() * (o.length - 1))
    ;[o[i], o[j]] = [o[j], o[i]]
    
    return o
  }
  
  // Tournament selection
  function tournament(pop: number[][], fitnesses: number[]): number[] {
    let best = Math.floor(Math.random() * pop.length)
    for (let i = 1; i < TOURNAMENT_SIZE; i++) {
      const candidate = Math.floor(Math.random() * pop.length)
      if (fitnesses[candidate] < fitnesses[best]) {
        best = candidate
      }
    }
    return pop[best].slice()
  }
  
  // Initialize population with some heuristic solutions
  let pop = Array.from({ length: POP }, randOrder)
  
  // Seed with nearest neighbor solution for better starting point
  pop[0] = nearestNeighbor(m).order
  
  // For larger problems, seed a few more with greedy solutions
  if (n > 30 && POP > 10) {
    for (let i = 1; i < Math.min(5, POP); i++) {
      pop[i] = nearestNeighbor(m).order // Multiple NN runs can give different results
    }
  }
  
  let bestEver: number[] | null = null
  let bestEverFitness = Infinity
  let stagnantGens = 0
  const MAX_STAGNANT = n > 50 ? 15 : 25 // Early stopping for large problems
  
  for (let g = 0; g < GEN; g++) {
    // Calculate all fitnesses once
    const fitnesses = pop.map(fitness)
    
    // Track best solution
    const genBestIdx = fitnesses.indexOf(Math.min(...fitnesses))
    if (fitnesses[genBestIdx] < bestEverFitness) {
      bestEverFitness = fitnesses[genBestIdx]
      bestEver = pop[genBestIdx].slice()
      stagnantGens = 0
    } else {
      stagnantGens++
    }
    
    // Early stopping if no improvement for too long
    if (stagnantGens > MAX_STAGNANT) {
      break
    }
    
    // Sort population by fitness for elitism
    const indexed = pop.map((ind, i) => ({ ind, fit: fitnesses[i] }))
    indexed.sort((a, b) => a.fit - b.fit)
    pop = indexed.map(x => x.ind)
    
    // Keep elite
    const elite = pop.slice(0, ELITE_SIZE)
    const newPop: number[][] = [...elite]
    
    // Generate new population
    while (newPop.length < POP) {
      let child: number[]
      
      if (Math.random() < CROSS_RATE) {
        // Crossover
        const p1 = tournament(pop, fitnesses)
        const p2 = tournament(pop, fitnesses)
        child = orderCrossover(p1, p2)
        
        // Apply mutation (mix of inversion and swap)
        if (Math.random() < 0.5) {
          child = inversionMutate(child)
        } else {
          child = swapMutate(child)
        }
      } else {
        // Just mutate a selected parent
        const parent = tournament(pop, fitnesses)
        child = inversionMutate(parent)
      }
      
      // Validate before adding
      if (isValidTour(child)) {
        newPop.push(child)
      } else {
        // If invalid, use a random valid tour instead
        newPop.push(randOrder())
      }
    }
    
    pop = newPop
    
    // Increase mutation rate if stagnant
    if (stagnantGens > 20 && Math.random() < 0.3) {
      // Introduce random individual for diversity
      pop[pop.length - 1] = randOrder()
    }
  }
  
  // Final sort to get best
  pop.sort((a, b) => fitness(a) - fitness(b))
  const order = bestEver || pop[0]
  const t = totalForOrder(order, m)
  return { order, totalDistance: t.dist, totalDuration: t.dur, legs: t.legs }
}

export function computeTSP(algo: AlgoKey, m: Matrix, extra?: { options?: GeneticOptions }): RouteResult | null {
  // Validate matrix structure
  if (!m || !m.distances || !m.durations) {
    throw new Error('Invalid matrix: missing distances or durations')
  }
  
  const n = m.distances.length
  if (n === 0) {
    throw new Error('Invalid matrix: empty distances array')
  }
  
  if (m.durations.length !== n) {
    throw new Error(`Invalid matrix: distances length (${n}) != durations length (${m.durations.length})`)
  }
  
  // Validate matrix is square
  for (let i = 0; i < n; i++) {
    if (!m.distances[i] || m.distances[i].length !== n) {
      throw new Error(`Invalid matrix: row ${i} has incorrect length`)
    }
    if (!m.durations[i] || m.durations[i].length !== n) {
      throw new Error(`Invalid matrix: durations row ${i} has incorrect length`)
    }
  }
  
  switch (algo) {
    case 'brute_force':
      if (m.distances.length > 11) {
        // safety: brute force explodes factorially; skip for large sets
        // O(n!) - 12! = 479,001,600 permutations is too much
        return null
      }
      return bruteForce(m)
    case 'nearest_neighbor':
      return nearestNeighbor(m)
    case 'genetic':
      return genetic(m, extra?.options)
  }
}

File: backend/app/tsp.py
from typing import List, Tuple, Dict, Any

Matrix = Dict[str, List[List[float]]]

def total_for_order(order: List[int], m: Matrix) -> Dict[str, Any]:
    legs = []
    total_d, total_t = 0.0, 0.0
    for a, b in zip(order[:-1], order[1:]):
        d = m['distances'][a][b]
        t = m['durations'][a][b]
        legs.append({"from": a, "to": b, "distance": d, "duration": t})
        total_d += d
        total_t += t
    return {"order": order, "totalDistance": total_d, "totalDuration": total_t, "legs": legs}

# Algorithms

def brute_force(m: Matrix) -> Dict[str, Any] | None:
    import itertools
    n = len(m['durations'])
    if n > 11:
        return None  # safety
    best = None
    for perm in itertools.permutations(range(1, n)):
        order = [0, *perm]
        res = total_for_order(order, m)
        if best is None or res['totalDuration'] < best['totalDuration']:
            best = res
    return best

def nearest_neighbor(m: Matrix) -> Dict[str, Any]:
    n = len(m['durations'])
    unvisited = set(range(1, n))
    order = [0]
    cur = 0
    while unvisited:
        nxt = min(unvisited, key=lambda j: m['durations'][cur][j])
        unvisited.remove(nxt)
        order.append(nxt)
        cur = nxt
    return total_for_order(order, m)

def two_opt(m: Matrix) -> Dict[str, Any]:
    n = len(m['durations'])
    # start with NN
    best = nearest_neighbor(m)
    order = best['order']
    improved = True
    def delta(i, k, order):
        a, b = order[i-1], order[i]
        c, d = order[k], order[(k+1) % len(order)] if k+1 < len(order) else None
        if d is None:
            return 0
        return (m['durations'][a][c] + m['durations'][b][d]) - (m['durations'][a][b] + m['durations'][c][d])
    while improved:
        improved = False
        for i in range(1, n-2):
            for k in range(i+1, n-1):
                if delta(i, k, order) < 0:
                    order[i:k+1] = reversed(order[i:k+1])
                    improved = True
        
    return total_for_order(order, m)

def genetic(m: Matrix, generations: int = 200, pop_size: int = 64, mutation_rate: float = 0.1) -> Dict[str, Any]:
    import random
    
    durations = m["durations"]
    n = len(durations)

    # ---- Helpers ----

    def cost(order: List[int]) -> float:
        """Total duration of a route."""
        return total_for_order(order, m)["totalDuration"]

    def fitness(order: List[int]) -> float:
        """Higher fitness = shorter duration."""
        return 1.0 / (1e-9 + cost(order))

    def next_unvisited_in_parent(parent, current, visited):
        """SCX helper."""
        try:
            idx = parent.index(current)
        except ValueError:
            for c in parent:
                if c not in visited:
                    return c
            return None

        for offset in range(1, len(parent) + 1):
            cand = parent[(idx + offset) % len(parent)]
            if cand not in visited:
                return cand
        return None

    def scx(p1, p2):
        """Sequential Constructive Crossover."""
        child = [0]
        visited = {0}
        current = 0
        
        while len(child) < n:
            c1 = next_unvisited_in_parent(p1, current, visited)
            c2 = next_unvisited_in_parent(p2, current, visited)

            if c1 is None and c2 is None:
                # add remaining
                for c in p1:
                    if c not in visited:
                        child.append(c)
                break

            if c1 is None:
                chosen = c2
            elif c2 is None:
                chosen = c1
            else:
                d1 = durations[current][c1]
                d2 = durations[current][c2]
                chosen = c1 if d1 < d2 else c2

            child.append(chosen)
            visited.add(chosen)
            current = chosen

        return child

    def mutate(order):
        if random.random() < mutation_rate and len(order) > 3:
            i, j = random.sample(range(1, len(order)), 2)
            order[i], order[j] = order[j], order[i]

    # ---- initialize population ----
    base = list(range(1, n))
    pop = [[0, *random.sample(base, len(base))] for _ in range(pop_size)]

    # ---- GA loop ----
    for _ in range(generations):
        # Evaluate + sort (best = first)
        pop.sort(key=lambda o: cost(o))
        
        # Elitism: keep best 25%
        elite_size = max(2, pop_size // 4)
        elites = pop[:elite_size]

        new_pop = elites[:]
        while len(new_pop) < pop_size:
            p1, p2 = random.sample(elites, 2)
            child = scx(p1, p2)
            mutate(child)
            new_pop.append(child)

        pop = new_pop

    best = min(pop, key=cost)
    return total_for_order(best, m)

